### PR TITLE
Disable autodetect when setting is off / Improve the full-pipeline info

### PR DESCRIPTION
**Pull request by yhmtsai**
This PR disables the autodetct when settings are explicitly off and improves the full-pipeline information.

Disable autodetect:
When something is broken in the environment for some reason, disable the corresponding backend to avoid issues.

Full pipeline information:
Ginkgo currently uses two stages - quick and full to test PRs.
The PR is usually required to pass the full test before merging the PR.
When no run or skip the full pipeline, it marks the full pipeline failed and gives the short information.
The detail is forwarded to the job. The output of the job also describes how to run or skip the full pipeline.
image
When skipping the full pipeline by `1:ST:skip-full-test`, it marks the full pipeline successful and gives it is skipped.
The detail is still forwarded to the job.
image

Skipping full test has a higher priority than running full test

### CODE REVIEW

**Review by upsj — APPROVED** [review]
LGTM! We should also test this behavior - since it's only looking at the CMake output, it's basically free.

File: CMakeLists.txt [hunk 1]
```diff
@@ -14,7 +14,9 @@ set(Ginkgo_VERSION_TAG "develop")
 set(PROJECT_VERSION_TAG ${Ginkgo_VERSION_TAG})
 
 # Determine which modules can be compiled
-include(cmake/hip_path.cmake)
+if(NOT DEFINED GINKGO_BUILD_HIP OR GINKGO_BUILD_HIP)
+    include(cmake/hip_path.cmake)
+endif()
```
Comment (upsj) [inline CMakeLists.txt:19]: is there any harm to including this every time?

File: cmake/autodetect_executors.cmake [hunk 2]
```diff
@@ -3,45 +3,46 @@ set(GINKGO_HAS_MPI OFF)
...
 include(CheckLanguage)
-check_language(CUDA)
-try_compile(GKO_CAN_COMPILE_DPCPP ${PROJECT_BINARY_DIR}/dpcpp
-    SOURCES ${PROJECT_SOURCE_DIR}/dpcpp/test_dpcpp.dp.cpp
-    CXX_STANDARD 17)
 
-if(OpenMP_CXX_FOUND)
-    if(NOT DEFINED GINKGO_BUILD_OMP)
+if (NOT DEFINED GINKGO_BUILD_OMP)
+    find_package(OpenMP 3.0)
```
Comment (tcojean) [inline cmake/autodetect_executors.cmake:10]: I think CUDA also needs this `find_package`

**Review by tcojean — APPROVED** [review]
LGTM

File: cmake/autodetect_executors.cmake (hunk 2 above)
Comment (yhmtsai) [reply cmake/autodetect_executors.cmake:10]: Is it done by check_language(CUDA)?

File: CMakeLists.txt (hunk 1 above)
Comment (yhmtsai) [reply CMakeLists.txt:19]: no, it avoids introducing many unrelated hip into cmake

Comment (upsj) [reply CMakeLists.txt:19]: I guess it's okay for now, we should probably move to native CMake support for HIP in the near future, anyways.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -247,21 +247,35 @@ trigger_pipeline:
...
+        ENABLE_FULL_PIPELINE=$(echo "${PR_CONTENT}" | jq -r \
           'any( [.labels | .[] | .name ] | .[] ; . == "1:ST:ready-to-merge" or . == "1:ST:run-full-test")')
-        if [[ "${ENABLE_FULL_PIPELINE}" == "true" ]]; then
-          echo "trigger full pipeline"
+        if [[ "${SKIP_FULL_PIPELINE}" == true ]]; then
+          echo "Skipping the full pipeline and making the full pipeline success"
+          curl -X POST -H "Accept: application/vnd.github.v3+json" -H "Authorization: token ${BOT_STATUS_TOKEN}" \
+            https://api.github.com/repos/ginkgo-project/ginkgo/statuses/${CI_COMMIT_SHA} \
+            -d "{\"state\":\"success\",\"context\":\"ci/gitlab/full\",\"target_url\":\"${CI_JOB_URL}\",\"description\":\"skip full pipeline\"}"
+          echo "If want to run full test, remove 1:ST:skip-full-test and add 1:ST:ready-to-merge or 1:ST:run-full-test. Rerun this job or have a new pipeline."
```
Comment (upsj) [inline .gitlab-ci.yml:261]:
```suggestion
          echo "If want to run full test, remove 1:ST:skip-full-test and add 1:ST:ready-to-merge or 1:ST:run-full-test. Rerun this job or start a new pipeline."
```

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -247,21 +247,35 @@ trigger_pipeline:
...
+        echo "Checking whether the PR contains 1:ST:ready-to-merge, 1:ST:run-full-test, or 1:ST:skip-full-test labels"
+        PR_CONTENT=$(curl -X GET -s -H "Accept: application/vnd.github.v3+json" -H "Authorization: token ${BOT_STATUS_TOKEN}" \
+          "https://api.github.com/repos/ginkgo-project/ginkgo/issues/${PR_ID}")
+        SKIP_FULL_PIPELINE=$(echo "${PR_CONTENT}" | jq -r 'any( [.labels | .[] | .name ] | .[] ; . == "1:ST:skip-full-test")')
+        ENABLE_FULL_PIPELINE=$(echo "${PR_CONTENT}" | jq -r \
           'any( [.labels | .[] | .name ] | .[] ; . == "1:ST:ready-to-merge" or . == "1:ST:run-full-test")')
-        if [[ "${ENABLE_FULL_PIPELINE}" == "true" ]]; then
-          echo "trigger full pipeline"
+        if [[ "${SKIP_FULL_PIPELINE}" == true ]]; then
+          echo "Skipping the full pipeline and making the full pipeline success"
```
Comment (upsj) [inline .gitlab-ci.yml:257]:
```suggestion
          echo "Skipping the full pipeline and making the full pipeline succeed"
```

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -247,21 +247,35 @@ trigger_pipeline:
...
             https://gitlab.com/api/v4/projects/6431537/trigger/pipeline
+          echo "If want to skip full pipeline, add 1:ST:skip-full-test. Rerun this job or have a new pipeline."
         else
-          echo "does not contain required labels"
+          echo "Not containing required labels and making the full pipeline fail."
+          curl -X POST -H "Accept: application/vnd.github.v3+json" -H "Authorization: token ${BOT_STATUS_TOKEN}" \
+            https://api.github.com/repos/ginkgo-project/ginkgo/statuses/${CI_COMMIT_SHA} \
+            -d "{\"state\":\"failure\",\"context\":\"ci/gitlab/full\",\"target_url\":\"${CI_JOB_URL}\",\"description\":\"no run/skip full pipeline\"}"
+          echo "If the pr does not need to run full pipeline before merging, rerun this job or have a new pipeline for this pr after adding 1:ST:skip-full-test."
+          echo "If the pr need to run full pipeline before merging, need to rerun this job or have a new pipeline for this pr after adding 1:ST:run-full-test or 1:ST:ready-to-merge."
```
Comment (upsj) [inline .gitlab-ci.yml:273]:
```suggestion
          echo "If the pr does not need to run full pipeline before merging, rerun this job or start a new pipeline for this pr after adding 1:ST:skip-full-test."
          echo "If the pr needs to run full pipeline before merging, need to rerun this job or start a new pipeline for this pr after adding 1:ST:run-full-test or 1:ST:ready-to-merge."
```

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -247,21 +247,35 @@ trigger_pipeline:
...
+            -d "{\"state\":\"success\",\"context\":\"ci/gitlab/full\",\"target_url\":\"${CI_JOB_URL}\",\"description\":\"skip full pipeline\"}"
+          echo "If want to run full test, remove 1:ST:skip-full-test and add 1:ST:ready-to-merge or 1:ST:run-full-test. Rerun this job or have a new pipeline."
+        elif [[ "${ENABLE_FULL_PIPELINE}" == "true" ]]; then
+          echo "Running full pipeline."
           curl -X POST -F token=${CI_JOB_TOKEN} -F "ref=${CI_COMMIT_REF_NAME}" -F "variables[STATUS_CONTEXT]=full" \
             https://gitlab.com/api/v4/projects/6431537/trigger/pipeline
+          echo "If want to skip full pipeline, add 1:ST:skip-full-test. Rerun this job or have a new pipeline."
         else
-          echo "does not contain required labels"
+          echo "Not containing required labels and making the full pipeline fail."
```
Comment (upsj) [inline .gitlab-ci.yml:268]:
```suggestion
          echo "Didn't find required labels, so we're making the full pipeline fail."
```

**Review by upsj — APPROVED** [review]
LGTM! Just a few nits on wording

**Pull request merged**